### PR TITLE
Sort town resident list by date they joined it

### PR DESCRIPTION
**Pull request by Warriorrrr**
#### Description: 

Sorts resident lists by the date the residents joined the town at.
____
#### New Nodes/Commands/ConfigOptions: 

____
#### Relevant Towny Issue ticket:

____
- [x] I have tested this pull request for defects on a server. 
* Tested on flatfile, haven't tested SQL but should work fine.

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

File: src/com/palmergames/bukkit/towny/object/Resident.java [hunk 1]
```diff
@@ -378,6 +379,7 @@ public void setTown(Town town) throws AlreadyRegisteredException {
 		this.town = town;
 		updatePerms();
 		town.addResident(this);
+		setJoinedTownAt(System.currentTimeMillis());
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/Resident.java:382]: This method is called during loading so all residents would end up having their join dates set to the server load up.

File: src/com/palmergames/bukkit/towny/object/Resident.java [hunk 2]
```diff
@@ -945,6 +947,14 @@ public void save() {
 		TownyUniverse.getInstance().getDataSource().saveResident(this);
 	}
 
+	public long getJoinedTownAt() {
+		return joinedTownAt;
+	}
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/Resident.java:952]:
```suggestion
	public long getJoinedTownAt() {
        if (joinedTownAt == null)
            return 0;
		return joinedTownAt;
	}
```

In order to handle players who joined the town prior to this update.

**Review by LlmDl — CHANGES_REQUESTED** [review]
On top of the things I've pointed out, the resident columns in the SQLSchema will have to have the new column added to it. 

The resident loading will probably need a loading-only method, that or the join date will have to be set in the town-add-resident code after the resident is set to the Town. 

I hadn't realized this list would no longer be in order, but I believe that since the change made to not save the resident list in the town this list is no longer in order of join date and rather is probably more alphabetical as residents are loaded in and have their town assigned.

**Review by LlmDl — APPROVED** [review]

File: src/com/palmergames/bukkit/towny/object/Town.java [hunk 3]
```diff
@@ -228,7 +230,7 @@ public void setNation(Nation nation) throws AlreadyRegisteredException {
 
 	@Override
 	public List<Resident> getResidents() {
-		return Collections.unmodifiableList(residents);
+		return Collections.unmodifiableList(residents.stream().sorted(Comparator.comparingLong(Resident::getJoinedTownAt)).collect(Collectors.toList()));
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/object/Town.java:233]: I would actually make a separate method that returns a sorted get residents. The general `getResidents()` gets called for a lot of things where it does not need to be sorted, so the sorting is just an unnecessary perf hit.

Comment (Warriorrrr) [reply src/com/palmergames/bukkit/towny/object/Town.java]: Adressed the performance aspect of it in 27ac19bcb62dd2a4ef3af35941711cedc9035aaa, residents will now only get sorted once at startup, which removes the need for a 2nd method.

File: src/com/palmergames/bukkit/towny/object/Town.java [hunk 4]
```diff
@@ -1383,4 +1402,18 @@ public UUID getUuid() {
...
+	public long getJoinedNationAt() {
+		return joinedNationAt;
+	}
+
+	public void setJoinedNationAt(long joinedNationAt) {
+		this.joinedNationAt = joinedNationAt;
+	}
+
+	private void sortResidents() {
+		List<Resident> sortedResidents = new ArrayList<Resident>(residents.stream().sorted(Comparator.comparingLong(Resident::getJoinedTownAt)).collect(Collectors.toList()));
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/object/Town.java:1414]: Don't need the `new ArrayList<Resident>()` part, the stream will collect it to a list already.

File: src/com/palmergames/bukkit/towny/object/TownBlock.java [hunk 5]
```diff
@@ -418,4 +426,12 @@ public void setPlotObjectGroup(PlotGroup group) {
 	public void save() {
 		TownyUniverse.getInstance().getDataSource().saveTownBlock(this);
 	}
+
+	public Long getClaimedAt() {
```
Comment (silverwolfg11) [inline src/com/palmergames/bukkit/towny/object/TownBlock.java:430]: Unnecessary boxing (`long)`

**Review by silverwolfg11 — APPROVED** [review]

**Pull request merged**